src/Chapter3.hs

```diff
@@ -479,7 +528,53 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
-
+data City = City
+    { cityCastle :: FortifiedBuilding
+    , citySpecialBuilding :: SpecialBuilding
+    , cityHouses :: [House]
+    } deriving (Show, Eq)
+
+data FortifiedBuilding 
+    = None 
+    | Castle String 
+    | CastleWithWalls String deriving (Show, Eq)
+
+data SpecialBuilding = Church | Library deriving (Show, Eq)
+
+data House = One | Two | Three | Four deriving (Show, Eq)
+
+buildCastle :: City -> String -> City
+buildCastle city name = city {cityCastle = createdCastle}
+    where createdCastle = cityCastle city `recreateCastleWith` name
+
+recreateCastleWith :: FortifiedBuilding -> String -> FortifiedBuilding
+recreateCastleWith building name = case building of
+    CastleWithWalls _ -> CastleWithWalls name
+    _ -> Castle name
+ 
+buildHouse :: City -> House -> City
+buildHouse city house = city {cityHouses = house : cityHouses city}
+
+countPeople :: House -> Int
+countPeople house = case house of
+    One   -> 1
+    Two   -> 2
+    Three -> 3
+    Four  -> 4
+    
+hasAtLeastNPeople :: City -> Int -> Bool
+hasAtLeastNPeople city n = cityHouses city `hasPeople` 0 
+    where 
+      hasPeople :: [House] -> Int -> Bool 
+      hasPeople [] _ = False    
+      hasPeople (x:xs) acc = 
+        let newAcc = acc + countPeople x
+        in if newAcc >= n then True else xs `hasPeople` newAcc  
```

Comment:
Neat! In this case, you would check only the necessary amount of houses, clever 👍🏼